include/ensmallen_bits/moead/moead_impl.hpp

```diff
@@ -0,0 +1,369 @@
+      // 2.4 Update of the neighbouring solutions.
+      for (size_t idx = 0;idx < neighbourhoodSize;++idx)
+      {
+        if (DecomposedSingleObjective(weights.col(weightNeighbourIndices(i, idx)),
+              idealPoint.col(0), evaluatedCandidate.col(0))
+            <= DecomposedSingleObjective(
+               weights.col(weightNeighbourIndices(i,idx)),
+               idealPoint.col(0), FValue.col(weightNeighbourIndices(i, idx))))
+        {
+          population.at(weightNeighbourIndices(i, idx)) = candidate[0];
+          FValue.col(weightNeighbourIndices(i, idx)) = evaluatedCandidate.col(0);
+        }
+      }
+
+      // 2.5 Updating External Population.
+      if (!externalPopulation.empty())
+      {
+        arma::mat first(numObjectives, 1);
+        auto df = [&](MatType firstMat) -> bool
+        {
+          std::vector<MatType> wrapperFirst(1), wrapperSecond(1);
+          wrapperFirst[0] = firstMat;
+          EvaluateObjectives(wrapperFirst, objectives, first);
+          return Dominates(evaluatedCandidate.col(0), first.col(0));
+        };
+        //! Remove the part that is dominated by candidate.
+        externalPopulation.erase(
+            std::remove_if(
+              externalPopulation.begin(),
+              externalPopulation.end(),
+              df), externalPopulation.end());
+
+        //! Check if any of the remaining members of external population dominate
+        //! candidate.
+        bool flag = 0;
+        std::vector<MatType> wrapperFirst(1);
+        for (size_t idx = 0; idx < externalPopulation.size(); ++idx)
+        {
+          wrapperFirst[0]=externalPopulation[idx];
+          first.clear();
+          first.resize(numObjectives, 1);
+          EvaluateObjectives(wrapperFirst, objectives, first);
+          if (Dominates(first.col(0), evaluatedCandidate.col(0)))
+          {
+            flag = 1;
+            break;
+          }
+        }
+        if (flag == 0)
+        {
+          externalPopulation.push_back(candidate[0]);
+          externalPopulationFValue.push_back(evaluatedCandidate.col(0));
+        }
+      }
+      else
+      {
+        externalPopulation.push_back(candidate[0]);
+        externalPopulationFValue.push_back(evaluatedCandidate.col(0));
+      }
+    }
+  }
+  bestFront = std::move(externalPopulation);
+
+  Callback::EndOptimization(*this, objectives, iterate, callbacks...);
+
+  ElemType performance = std::numeric_limits<ElemType>::max();
+  for (arma::Col<ElemType> objective: externalPopulationFValue)
+  {
+    if (arma::accu(objective) < performance)
+      performance = arma::accu(objective);
+  }
+
+  return performance;
+}
+
+//! Perform mutation of the candidate.
+template<typename MatType>
+inline void MOEAD::Mutate(MatType& child,
+    const double& rate,
+    const arma::vec& lowerBound,
+    const arma::vec& upperBound)
+{
+  size_t numVariables = lowerBound.n_elem;
+  double rnd, delta1, delta2, mutationPower, deltaq;
+  double y, yl, yu, val, xy;
```

Comment:
I was wondering if we could have a bit more descriptive names, if these are the symbols used in the paper, it is fine! :)

Reply:
Actually I did not see the paper first hand. I used [this](https://github.com/esa/pagmo/blob/master/src/algorithm/moea_d.cpp) implementation as a reference. I have made them more descriptive.

Reply:
Ah, I see.

Reply:
Ahh, that's why I had some issues reconstructing the mutation step from the paper.

Reply:
Yes. Will do.

Reply:
So what I found out is, `Mutate` implements polynomial mutation and the crossover step is just plain crossover, where each gene of the variable is chosen from one of the two randomly selected members. This isn't DE crossover at all. I should have gone through the paper earlier but could not get my hands on it. So should I implement the crossover step as DE crossover or just remove it from the citation and docs? Sorry for the mess up.:bow: